Assert metadata schema location (#6516)

dsotirho-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=anvilprod-promotion.md`,
`&template=prod-promotion.md`, `&template=anvilprod-hotfix.md`, `&template=prod-
hotfix.md`, `&template=backport.md` or `&template=upgrade.md` to switch the
template.
-->

Connected issues: #6516

## Notes for operator

- [x] Full reindex of `dev`
- [x] Reindex of LungMAP catalogs in `prod`


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or [comment](https://github.com/DataBiosphere/azul/pull/6747#issuecomment-2593492604) in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>


### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator


### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>


### System administrator

- [x] Background migrations for [`dev.gitlab`](https://gitlab.dev.singlecell.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for [`anvildev.gitlab`](https://gitlab.anvil.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`


### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>


### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

dsotirho-ucsc committed:
[2/2] Assert metadata schema location (#6516)

dsotirho-ucsc commented:
[6747_IT_2024-12-09.txt](https://github.com/user-attachments/files/18069699/6747_IT_2024-12-09.txt)

achave11-ucsc reviewed (commented):
IIRC, we prefer committing test coverage first, followed by source code changes in order to demonstrate effectiveness via build status checks. Consider inverting the order of your commits. Also, I'd keep the commits simple and straight forward, a commit tests the feature the other is the feature. And lastly, for my personal preference, the patches are a bit duplicative, consider writing a handler for patching, and to keep the call sites concise.

dsotirho-ucsc committed:
[1/2] Assert metadata schema location (#6516)

Add tests

dsotirho-ucsc commented:
[6747_IT_2024-12-12.txt](https://github.com/user-attachments/files/18116368/6747_IT_2024-12-12.txt)

achave11-ucsc reviewed (dismissed):
LGTM ✅

hannes-ucsc commented on src/humancellatlas/data/metadata/api.py:
```diff
@@ -110,12 +118,14 @@ def from_json(cls,
-        schema_name = described_by.rpartition('/')[2]
+        described_by = furl(content['describedBy'])
+        require(described_by.netloc in valid_schema_domains,
+                described_by.netloc)
```
```suggestion
                'Invalid schema domain', netloc, valid_schema_domains)
```

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -85,6 +95,7 @@ def _rename_keys(self, d, **kwargs):
                 d[new_name] = d[old_name]
                 del d[old_name]
 
+    @valid_schema_domains(['schema.dev.data.humancellatlas.org'])
```
This is a marginal aspect of our tests and I don't think I want to pollute this many test methods with a decorator invocation for the sake of each test having a minimal set of valid domains. We should provide this patch in a super class of our unit tests case classes and just allow the union of all observed domains there.

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +778,57 @@ def test_name_substitution(self):
+        json_body = self.json_body(domain='schema.humancellatlas.org')
+        with self.assertRaises(RequirementError):
+            Protocol.from_json(json_body)
+
```
Since you are actively working on #6666, let's not have a FIXME and just add this case in that PR.

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +778,57 @@ def test_name_substitution(self):
+
+    def test_valid_schema_domain(self):
+        json_body = self.json_body(domain='schema.humancellatlas.org')
+        file = Protocol.from_json(json_body)
```
Why is the variable called `file` when it's referring to an instance of `Protocol`?

dsotirho-ucsc replied:
The variable name was a oversight from an earlier version that was using a `File` schema

dsotirho-ucsc commented:
[6747_IT_2024-12-13.txt](https://github.com/user-attachments/files/18133094/6747_IT_2024-12-13.txt)

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +760,34 @@ def test_name_substitution(self):
+                'schema_type': 'protocol'
+            },
+            'hca_ingest': {
+                'document_id': str(uuid4()),
```
```suggestion
                'document_id': str(uuid4()),
```

Why did you decide not to hard-code the UUID?

dsotirho-ucsc replied:
To make it clear that this is a test entity, not from any canned/live bundle.

hannes-ucsc replied:
I wouldn't see that and come to the conclusion you intended.

We want our tests to be reproducible. Generating a v4 UUID adds randomness. 

Even if we use randomness, say, in the IT, we log the random generator's seed so that we can still reproduce a test, despite the randomness. Besides, if I wanted to check if the hard-coded value relates to a can, I can always do a text search. And we never tracked whether a UUID relates to a live bundle. Most cans come from live bundles, but then we modified the cans so I wouldn't be surprised to see a few new UUIDs in them, either.

Please allocate a new UUID and hard-code it here.

hannes-ucsc reviewed (changes requested):
https://github.com/DataBiosphere/azul/pull/6747#discussion_r1887778452

dsotirho-ucsc commented:
[6747_IT_2024-12-17.txt](https://github.com/user-attachments/files/18169522/6747_IT_2024-12-17.txt)

hannes-ucsc reviewed (changes requested):
```diff
Index: src/humancellatlas/data/metadata/api.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/humancellatlas/data/metadata/api.py b/src/humancellatlas/data/metadata/api.py
--- a/src/humancellatlas/data/metadata/api.py	(revision 763233ff2353615be1360de78b30aaf760b283a9)
+++ b/src/humancellatlas/data/metadata/api.py	(date 1734489196187)
@@ -42,7 +42,6 @@
     adict,
     dict_merge,
 )
-
 from azul.indexer.document import (
     EntityReference,
 )
@@ -120,12 +119,12 @@
         content = json.get('content', json)
         described_by = furl(content['describedBy'])
         require(described_by.netloc in valid_schema_domains,
-                'Invalid schema domain', described_by.netloc, valid_schema_domains)
+                'Unexpected schema domain', described_by, valid_schema_domains)
         schema_name = described_by.path.segments[-1]
         try:
             sub_cls = entity_types[schema_name]
         except KeyError:
-            raise TypeLookupError(str(described_by))
+            raise TypeLookupError(described_by)
         return sub_cls(json, **kwargs)
 
     def __init__(self,
@@ -167,7 +166,7 @@
 
 class TypeLookupError(Exception):
 
-    def __init__(self, described_by: str) -> None:
+    def __init__(self, described_by: furl) -> None:
         super().__init__(f"No entity type for schema URL '{described_by}'")
 
 
Index: test/azul_test_case.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/azul_test_case.py b/test/azul_test_case.py
--- a/test/azul_test_case.py	(revision 763233ff2353615be1360de78b30aaf760b283a9)
+++ b/test/azul_test_case.py	(date 1734488546770)
@@ -373,8 +373,10 @@
             'schema.staging.data.humancellatlas.org',
             'schema.integration.data.humancellatlas.org',
         ]
-        cls.addClassPatch(patch('humancellatlas.data.metadata.api.valid_schema_domains',
-                                domains))
+        from humancellatlas.data.metadata import api
+        cls.addClassPatch(patch.object(target=api,
+                                       attribute='valid_schema_domains',
+                                       new=domains))
 
 
 class CatalogTestCase(AzulUnitTestCase, metaclass=ABCMeta):
```

I don't think separating out the middle commit adds any clarity, especially since it adds the module variable to code under test without doing anything with it, other than patching it. I think you should do a split commit with test and fix being the two halves.

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +759,34 @@ def test_name_substitution(self):
+
+    def test_invalid_schema_domain(self):
+        json_body = self.json_body(domain='foo.humancellatlas.org')
+        with self.assertRaises(RequirementError):
```
This should assert more about the exception.

dsotirho-ucsc commented:
[6747_IT_2024-12-19.txt](https://github.com/user-attachments/files/18204287/6747_IT_2024-12-19.txt)

hannes-ucsc reviewed (changes requested):
There are two more places where we would expect the `describedBy` property: `links` and the `descriptor`. `links` is passed to the Bundle constructor. And I believe the `descriptor` is available via `ManifestEntry.json` but I could be wrong. Please confirm and request a PL slot. We may want to place the schema validation into the repository plugin or finally get rid of the ManifestEntry class and replace it with Descriptor so that the API matches what TDR gives us, instead of what the DSS used to give us, years ago.

Note that the ticket mentions the descriptor and what you've implemented here does not check the descriptor schema, AFAICT. We've both lost sight of that aspect.

dsotirho-ucsc commented:
[6747_IT_2025-01-13.txt](https://github.com/user-attachments/files/18400489/6747_IT_2025-01-13.txt)

hannes-ucsc reviewed (commented):
I'd like @nadove-ucsc's eyes on this, too. Please request a thorough peer review from them.

nadove-ucsc commented on src/azul/plugins/repository/tdr_hca/__init__.py:
```diff
@@ -189,6 +189,10 @@ def add_entity(self,
             descriptor = json.loads(row['descriptor'])
+            # FIXME Move validation of descriptor to the metadata API
+            #       https://github.com/DataBiosphere/azul/issues/6299
+            from humancellatlas.data.metadata.api import Entity  # noqa AZL131
```
Why is `noqa` used here instead of just properly formatting the import statement?

dsotirho-ucsc replied:
I originally thought the local import looked cleaner using one line rather than three.

nadove-ucsc reviewed (commented):
Code changes look good, but I'm confused regarding the PR's context. This doesn't appear to have anything to do with the `drs_uri` property, so the PR title makes no sense to me. Is the PR partial? Has the scope of the issue changed?

Consider whether it would be more appropriate to rename the PR to match the commit title and add a comment explaining why the PR title differs from the issue title.

dsotirho-ucsc commented:
PR https://github.com/HumanCellAtlas/metadata-schema/pull/1575 added the `drs_uri` property to the file descriptor schema.
This PR (https://github.com/DataBiosphere/azul/pull/6747) makes it so that all HCA schema URLs are required to use netloc `schema.humancellatlas.org` as requested in [comment](https://github.com/DataBiosphere/azul/issues/6516#issuecomment-2428334110).

dsotirho-ucsc commented:
[6747_IT_2025-01-15.txt](https://github.com/user-attachments/files/18428167/6747_IT_2025-01-15.txt)

hannes-ucsc commented on test/indexer/data/1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.tdr.hca.json:
```diff
@@ -7,8 +7,8 @@
             "version": "2019-09-24T09:35:06.958773Z",
-            "crc32c": "",
-            "sha256": "",
+            "crc32c": "00000000",
```
These changes should be in a separate, non-split commit. We'd like to keep them, even if we revert the changes that assert the schema location.

hannes-ucsc replied:
By these changes, I mean all changes to cans.

dsotirho-ucsc committed:
Add dummy checksums and valid describedBy values to canned TDR HCA bundle

dsotirho-ucsc commented:
[6747_IT_2025-01-21.txt](https://github.com/user-attachments/files/18499266/6747_IT_2025-01-21.txt)

dsotirho-ucsc commented:
@hannes-ucsc Note that I had to add `raw.githubusercontent.com` as an allowed schema domain due to two lungmap snapshots.

```
SELECT descriptor FROM `datarepo-dev-2e9ef7fd.lungmap_dev_20037472ea1d4ddb9cd356a11a6f0f76__20220401_20220404_lm2.sequence_file`

{
  "content_type": "text/fastq",
  "crc32c": "7b7fd9e2",
  "describedBy": "https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/lungmap-phantom-files/json_schema/system/file_descriptor.json",
  "drs_uri": null,
  "file_id": "42005932-b85c-4f08-9bfc-26262efc013f",
  "file_name": "UCSD_snATAC/hLung_D139_3yr_snATAC_Rep1_XH180_R2.fastq.gz",
  "file_version": "2022-02-22T18:41:29.333827Z",
  "schema_type": "file_descriptor",
  "schema_version": "2.1.1",
  "sha1": "a5dcb97c8dbe3704c0488354a59093ce5830adeb",
  "sha256": "41cb756661fad55010750241645fe06add4cf7ccad03ad8185dc1299f56e396b",
  "size": 4719609223
}
...

SELECT descriptor FROM `datarepo-dev-d57fd0c5.lungmap_dev_f899709cae2c4bb988f0131142e6c7ec__20220401_20220629_lm2.sequence_file`

{
  "content_type": ".gz",
  "crc32c": "8e92f6af",
  "describedBy": "https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/lungmap-phantom-files/json_schema/system/file_descriptor.json",
  "drs_uri": null,
  "file_id": "a8b6b9de-2038-4071-ab06-77038b737d97",
  "file_name": "copd1_periph_rep1_R1.fastq.gz",
  "file_version": "2022-03-03T19:02:53.832737Z",
  "schema_type": "file_descriptor",
  "schema_version": "2.1.1",
  "sha256": "04cfeac4c5258eb40f397fa272a0a0b25e3fa007d3977701f3f5326462ca011c",
  "size": 428420719
}
...
```

hannes-ucsc commented on src/humancellatlas/data/metadata/api.py:
```diff
@@ -97,6 +102,12 @@ def __init__(self, json: MutableJSON):
 
+valid_schema_domains = [
+    'schema.humancellatlas.org',
+    'raw.githubusercontent.com'  # for lungmap snapshots 2e9ef7fd & d57fd0c5
```
This is exactly the URL that we DO NOT want in any snapshot. That's why we created this PR. See my original [marching orders](https://github.com/DataBiosphere/azul/issues/6516#issuecomment-2428334110). We can't allow bad URLs to creep back into prod only because two snapshots in dev still have them. That's the wrong approach.

Please remove the offending snapshots from `dev` and create an issue about populating `dev` with updated snapshots from `prod`.

dsotirho-ucsc replied:
Created ticket #6841

dsotirho-ucsc committed:
[u] Remove LungMAP snapshots from dev with invalid schema locations

These two snapshots contain sequence file descriptors that use the schema:
https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/lungmap-phantom-files/json_schema/system/file_descriptor.json

dsotirho-ucsc committed:
[u] Remove dcp3 snapshohts from dev with invalid schema locations

This dcp3 snapshot contains project metadata that uses the schema:
https://schema.dev.data.humancellatlas.org/type/project/14.0.0/project

dsotirho-ucsc commented:
[6747_IT_2025-01-27.txt](https://github.com/user-attachments/files/18567858/6747_IT_2025-01-27.txt)

hannes-ucsc commented on deployments/dev/environment.py:
```diff
@@ -159,8 +159,9 @@ def mkdict(previous_catalog: dict[str, str],
-dcp3_sources = mkdict(dcp2_sources, 105, mkdelta([
+dcp3_sources = mkdict(dcp2_sources, 104, mkdelta([
     mksrc('bigquery', 'datarepo-dev-3d6d24ad', 'hca_dev_05657a599f9d4bb9b77b24be13aa5cea__20210827_20230215'),
+    mksrc('bigquery', 'datarepo-dev-76de829d', 'hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907', pop),
```
I only remember discussing the LunGMAP snapshots with you and the ticket you created only covers LunGMAP. Why is this one being removed?

dsotirho-ucsc replied:
This snapshot has a project entity that uses an invalid describedBy domain.

```
SELECT content FROM `datarepo-dev-76de829d.hca_dev_2043c65a1cf84828a6569e247d4e64f1__20210831_20210907.project`

{"describedBy":"https://schema.dev.data.humancellatlas.org/type/project/14.0.0/project","schema_type":"project","project_core":...

```

dsotirho-ucsc replied:
Created #6876

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +759,45 @@ def test_name_substitution(self):
+        expected = (
+            'Unexpected schema domain',
+            furl('https://foo.humancellatlas.org/type/protocol/7.1.0/protocol'),
+            [
+                'schema.humancellatlas.org',
+                'schema.dev.data.humancellatlas.org',
+                'schema.staging.data.humancellatlas.org',
+                'schema.integration.data.humancellatlas.org'
+            ]
```
Don't duplicate this. Define a class attribute and use it here.

hannes-ucsc commented on test/hca_metadata_api/test.py:
```diff
@@ -754,6 +759,45 @@ def test_name_substitution(self):
 
+class TestSchema(AzulUnitTestCase):
+
+    def json_body(self, domain: str) -> JSON:
```
```suggestion
    def _entity_json(self, schema_domain: str) -> JSON:
```

dsotirho-ucsc commented:
[6747_IT_2025-01-30.txt](https://github.com/user-attachments/files/18612180/6747_IT_2025-01-30.txt)

dsotirho-ucsc commented:
[6747_IT_2025-02-03.txt](https://github.com/user-attachments/files/18649270/6747_IT_2025-02-03.txt)

hannes-ucsc commented on src/azul/plugins/repository/tdr_hca/__init__.py:
```diff
@@ -189,6 +189,12 @@ def add_entity(self,
             self.stitched.add(entity.entity_id)
         if entity.entity_type.endswith('_file'):
             descriptor = json.loads(row['descriptor'])
+            # FIXME Move validation of descriptor to the metadata API
```
```suggestion
            # FIXME: Move validation of descriptor to the metadata API
```

hannes-ucsc commented on test/azul_test_case.py:
```diff
@@ -364,6 +365,22 @@ def _patch_lambda_env(cls):
+
+    @classmethod
+    def _patch_valid_schema_domains(cls):
+        from humancellatlas.data.metadata import (
+            api,
+        )
```
What's the reason for making this a local import?

dsotirho-ucsc replied:
I believe we added this as a local import during our investigation in PL how validation cold be added to `links` and `descriptor`. I might of missed that I was supposed un-localize the import when cleaning up our rough draft.
I've added a fixup commit to get rid of the local import.

hannes-ucsc replied:
> I've added a fixup commit to get rid of the local import.

Local import is still there.

dsotirho-ucsc committed:
Add FIXME (#6299)

dsotirho-ucsc commented:
[6747_IT_2025-02-05.txt](https://github.com/user-attachments/files/18678252/6747_IT_2025-02-05.txt)

dsotirho-ucsc commented:
[6747_IT_2025-02-06.txt](https://github.com/user-attachments/files/18697915/6747_IT_2025-02-06.txt)

hannes-ucsc commented:
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below
